research/conditional/utils/model_utils.py

```diff
@@ -49,90 +48,96 @@
+        return partial(
+            chungized_llm_loss_and_backward_pass, n_chungs=loss_checkpoint_chungs
+        )
 
 
-def chungized_llm_loss(
+def chungized_llm_loss_and_backward_pass(
     batch: LLMBatch,
     model: torch.nn.Module,
     mixed_precision: bool,
     vocab_size: int,
+    scaler: torch.cuda.amp.GradScaler,
     n_chungs: int,
+    backward_pass: bool,
+    num_accumulated_batches: int,
 ):
+    if isinstance(model, DDP):
+        model = model.module
     input_tokens = batch.input_ids
     gt_tokens = batch.target_ids
     mask = batch.should_calculate_loss
-
-    def make_custom_forward():
-        def custom_forward(*inputs):
-            output = model.head(inputs[0])
-            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
-                gt = inputs[1]
-                mask = inputs[2]
-                gt = gt.to(output.device)
-                loss = F.cross_entropy(
-                    output.reshape(-1, vocab_size),
-                    gt.reshape(-1).long(),
-                    reduction="none",
-                )
-
-                correct_tokens = gt.long() == output.argmax(dim=-1)
-                correct_tokens = correct_tokens.long().reshape(-1) * mask.reshape(-1)
-                correct_tokens = correct_tokens.sum()
-
-                total_tokens = mask.sum()
-
-            return loss[mask.reshape(-1) == 1], correct_tokens, total_tokens
-
-        return custom_forward
+    num_masked_tokens = mask.sum()
 
     with torch.autocast(
         device_type="cuda", enabled=mixed_precision, dtype=torch.float16
     ):
-        encoder_output = model.encoder(input_tokens)
-        chunged_inputs = torch.chunk(encoder_output, n_chungs, dim=0)
+        encoder_output: torch.Tensor = model.encoder(input_tokens)
+        encoder_output_det = encoder_output.detach()
+        if backward_pass:
+            encoder_output_det.requires_grad = True
+        chunged_inputs = torch.chunk(encoder_output_det, n_chungs, dim=0)
         chunged_non_masked_inputs = torch.chunk(gt_tokens, n_chungs, dim=0)
         chunged_non_masked_masks = torch.chunk(mask, n_chungs, dim=0)
 
-        num_tokens = 0
         total_loss = 0
         total_correct_tokens = 0
-        total_masked_tokens = 0
         for chunged_input, chunged_gt, chunged_mask in zip(
             chunged_inputs, chunged_non_masked_inputs, chunged_non_masked_masks
         ):
-            (
-                partial_loss_output,
-                partial_correct_tokens,
-                partial_masked_tokens,
-            ) = checkpoint(
-                make_custom_forward(), chunged_input, chunged_gt, chunged_mask
-            )
-            num_tokens += partial_loss_output.shape[0]
-            total_loss += partial_loss_output.sum()
+            output = model.head(chunged_input)
+            with torch.autocast(device_type="cuda", enabled=False, dtype=torch.float16):
```

Comment:
isn't this eqivalent to just deleting this context manager? since it's hardcoded to False

Reply:
Same as above, this part needs to have it turned off because of some autocast fuckery which slows it down if it is turned on

Reply:
noted!